Features: Use Autotools for building binary

teoching0705 opened the pull request:
These are the autotools related files aim to fix #10, Please checkout branch [autotools](https://github.com/ePSIC-DLS/mib2h5/tree/autotools) to see the working example.

Below are the instructions for the branch [autotools](https://github.com/ePSIC-DLS/mib2h5/tree/autotools):
Please install and build hdf5, c-blosc, and hdf5-blosc and then specify it using:

export HDFDIR=${path_to_hdf5_root_dir}
export BLOSCDIR=${path_to_c-blosc_root_dir}
export HBDIR=${path_to_hdf5-blosc_root_dir}

Then run autoreconf -i, you then should see an executable ./configure
Please then run ./configure, the target binary will be built inside src, i.e. src/mib2h5

Please run the binary with ONE input .mib file, i.e.:

./src/mib2h5 ${.mib_file_path}

This will then create the resulting h5 file under /scratch/30day_temp (You can change this in src/main.c)

teoching0705 committed:
Add in autotools file

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,53 @@
+AC_INIT([mib2h5], [1.5.2025], [teo.ching@diamond.ac.uk])
```
Use semantic version, let start with `0.0.1`. And even for calendar versioning it rarely starts with the date, look up 'semantic or calendar version' if you want to understand more.

ptim0626 replied:
[The third argument is for bug report](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.72/autoconf.html#index-AC_005fINIT-2), although this is normally an email address, it can be [an URL to a bug management web page](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.72/autoconf.html#Initializing-configure), use the GitHub repo issue page for this.

ptim0626 replied:
Let use the fifth argument to provide the project URL, which is the GitHub repository. For the fourth argument I think it is fine to use `mib2h5` as a tarball name.

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,53 @@
+AC_INIT([mib2h5], [1.5.2025], [teo.ching@diamond.ac.uk])
+AM_INIT_AUTOMAKE([-Wall foreign -Werror])
```
You have [`foreign` option](https://www.gnu.org/software/automake/manual/automake.html#index-Strictness_002c-foreign) here (which I also prefer), so I guess we don't need [the files required by the GNU standard](https://www.gnu.org/software/automake/manual/automake.html#index-Strictness_002c-gnu)?

ptim0626 replied:
I suggest [specifying the tarball format and compression as here](https://www.gnu.org/software/automake/manual/automake.html#index-Option_002c-tar_002dv7). I think we can use the `tar-pax` format (seems to be defined in 2001 which is 'modern' enough) with `dist-xz`.

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,53 @@
+
+AC_ARG_ENABLE([asan],
+  AS_HELP_STRING([--enable-asan], [Enable AddressSanitizer build]),
+  [asan_build=$enableval], [asan_build=no])
```
The debug build should not be the default, i.e. users won't want to type `--enable-debug=no` when they are using the package (release build). The debug build is more for developers' use, same for asan.

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,53 @@
+AM_INIT_AUTOMAKE([-Wall foreign -Werror])
+AC_PROG_CC
+AC_CONFIG_HEADERS([config.h])
+AC_CONFIG_FILES([Makefile src/Makefile])
```
The [suggested layout](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.72/autoconf.html#Autoconf-Input-Layout) put `AC_CONFIG_FILES` right before `AC_OUTPUT`, although I am not sure what differences it will have in our case, let follow what it is suggested.

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,53 @@
+# --- External paths (can be passed as environment variables) ---
+AC_ARG_VAR([HDFDIR], [Path to HDF5 installation])
+AC_ARG_VAR([BLOSCDIR], [Path to c-blosc installation])
+AC_ARG_VAR([HBDIR], [Path to hdf5-blosc installation])
```
From here we assume the hdf5, blosc and hdf5-blosc libraries are mandatory. While the hdf5 library is essential (the tool is to convert .mib to hdf5!), blosc-related libraries should not be an essential requirement.

One might think, given the advantages of blosc compression, everyone would like to use it. However, one reason people will try to avoid blosc compression is for downstream application. Perhaps the downstream application/analysis pipeline doesn't support opening hdf5 files with blosc-compressed dataset, or they are working in a restricted environment that doesn't provide the blosc library for decompression (e.g HPC system where the central environment is usually managed by others).

Our approach should make the blosc compression functionalities as optional, only when the users specify something like `--with-blosc` (provided by `AC_ARG_WITH`) we will build with the necessary blosc-related libraries.

ptim0626 replied:
As the hdf5 library is an essential requirement, we should:
1. check if the user provides a custom installation path via `--with-hdf5`;
2. if not, see if the user provide the installation path by env var (let use `HDF5_DIR` for it instead of `HDFDIR`, same as `h5py`);
3. the last resort is to check if it is already visible (by testing the presence of `h5cc` for example)

ptim0626 commented:
And we should provide a `main` function as the entry point, I think it is sufficient for it to just print out the version set in `configure.ac` for now when executing `mib2h5`.

ptim0626 commented:
A suggestion: we could [enable silent rules for improved readability](https://www.gnu.org/software/automake/manual/html_node/Automake-Silent-Rules.html), by using `AM_SILENT_RULES([yes])`.

teoching0705 committed:
Add logics to let user provide path for hdf5, blosc, and hdf5-blosc

This will let users to provide custom path for these dependencies.
--with-hdf5:
  provided path: use this
  not provided: detect in system path, if no then abort
--with-compression:
  provided: indicate the need for checking blosc and hdf5-blosc
  not provided: not compiling with those libraries
--with-blosc, --with-hdf5-blosc:
  provided path: use this path for those libraries
  not provided: if --with-compression is stated, check system path, if no then abort

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,115 @@
+AC_INIT([mib2h5], [0.0.1], [https://github.com/ePSIC-DLS/mib2h5], [mib2h5],[https://github.com/ePSIC-DLS/mib2h5])
```
The third argument should be issue page address, and to be picky, the spacing.

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,115 @@
+AC_INIT([mib2h5], [0.0.1], [https://github.com/ePSIC-DLS/mib2h5], [mib2h5],[https://github.com/ePSIC-DLS/mib2h5])
+AM_INIT_AUTOMAKE([-Wall foreign -Werror tar-pax])
```
I think you also need to add [`dist-xz`](https://www.gnu.org/software/automake/manual/automake.html#index-dist_002dxz-1).

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,115 @@
+# --- Build Type Options --
+AC_ARG_ENABLE([debug],
+  AS_HELP_STRING([--enable-debug], [Enable debug build]),
+  [debug_build=$enableval], [debug_build=yes])
```
Other than changing the message in help, you will need to change to `debug_build=no` for action-if-not-given

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,115 @@
+fi
+
+AC_PROG_CC
+AM_SILENT_RULES([yes])
```
This should go right after `AM_INIT_AUTOMAKE`. It is an initialisation step so this makes sense to be placed early in the script.

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,115 @@
+
+AC_PROG_CC
+AM_SILENT_RULES([yes])
+AC_CONFIG_HEADERS([config.h])
```
This is [one of the instantiating macros](https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Configuration-Headers.html) so place it early in the script.

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,115 @@
+  CFLAGS="$CFLAGS -Wall -Wextra -std=c11 -O2"
+fi
+
+AC_PROG_CC
```
This checks the C compiler and should be placed early as it performs a lot of tests and will affect other checks afterwards.

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,115 @@
+    ], [AC_MSG_ERROR([libhdf5 not found system-wide])])
+  ], [AC_MSG_ERROR([hdf5.h not found system-wide])])
+])
+
```
This doesn't check for environment variable defined by the user (common ones are  `HDF5_DIR`/`HDF5_ROOT`/`HDF5_HOME`) if user doesn't provide via `--with-hdf5` and before a system-wide check.

I suggest check for each commonly used environment variable one-by-one if `HDFDIR` is not defined, assign the corresponding one to `HDFDIR`, and set `CPPFLAGS` and `LDFLAGS` afterwards.

Some pseduo-code:
IF HDFDIR is defined (via --with-hdf5)
     using the user-specified path
ELSE IF the env var HDF5_DIR is defined
     set HDFDIR to the value of HDF5_DIR
ELSE IF the env var HDF5_ROOT is defined
     ...........
     
IF HDFDIR is defined
    set CPPFLAGS and LDFLAGS

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,115 @@
+  CFLAGS="$CFLAGS -Wall -Wextra -std=c11 -g -O0 -fanalyzer"
+else
+  CFLAGS="$CFLAGS -Wall -Wextra -std=c11 -O2"
+fi
```
Why not use the m4 macro `AS_IF` as earlier?

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,115 @@
+])
+
+# ---------------- Compression Option ----------------
+AC_ARG_WITH([compression],
+  [AS_HELP_STRING([--with-compression], [Enable compression with blosc and hdf5-blosc])],
+  [],
+  [with_compression=check])
```
Typically with `AC_ARG_WITH` the option is a path; use `AC_ARG_ENABLE` for a yes/no decision.

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,115 @@
+  # --- If user gave paths, use them directly ---
+  AS_IF([test -n "$BLOSCDIR" && test -n "$HBDIR"], [
+
+    CPPFLAGS="$CPPFLAGS -I$BLOSCDIR/include -I$HBDIR/src"
+    LDFLAGS="$LDFLAGS -L$BLOSCDIR/lib64 -L$HBDIR/build"
+    LDFLAGS="$LDFLAGS -Wl,-rpath=$BLOSCDIR/lib64 -Wl,-rpath=$HBDIR/build"
+
+    AC_CHECK_HEADER([blosc.h], [],
+      [AC_MSG_ERROR([blosc.h not found in $BLOSCDIR/include])])
+    AC_CHECK_LIB([blosc], [blosc_init], [],
+      [AC_MSG_ERROR([libblosc not found in $BLOSCDIR/lib64])])
+
+    AC_CHECK_HEADER([blosc_filter.h], [],
+      [AC_MSG_ERROR([blosc_filter.h not found in $HBDIR/src])])
+    AC_CHECK_LIB([blosc_filter], [register_blosc], [],
+      [AC_MSG_ERROR([libblosc_filter not found in $HBDIR/build])])
+
+    AC_DEFINE([HAVE_COMPRESSION], [1], [Define if compression is enabled])
+    AC_SUBST([LIBCOMPRESSION], ["-lblosc -lblosc_filter"])
+
+  ], [
+
+    # --- No user path: try system-wide ---
+    AC_CHECK_HEADER([blosc.h], [
+      AC_CHECK_LIB([blosc], [blosc_init], [BLOSC_OK=yes])
+    ])
+    AC_CHECK_HEADER([blosc_filter.h], [
+      AC_CHECK_LIB([blosc_filter], [register_blosc], [HDF5_BLOSC_OK=yes])
+    ])
+
+    AS_IF([test "x$BLOSC_OK" = xyes && test "x$HDF5_BLOSC_OK" = xyes], [
+      AC_DEFINE([HAVE_COMPRESSION], [1], [Define if compression is enabled])
+      AC_SUBST([LIBCOMPRESSION], ["-lblosc -lblosc_filter"])
+    ], [
+      AS_IF([test "x$with_compression" != xcheck], [
+        AC_MSG_FAILURE([Compression requested but blosc and hdf5-blosc not found system-wide, and no --with-blosc or --with-hdf5-blosc given])
+      ])
+    ])
+  ])
+])
```
Does this handle the case of enabling compression but with just one custom path (either blosc or hdf5-blosc)?

teoching0705 replied:
No, I think what this does is that check both dependencies if compression is enabled. 

I think there is no point of allowing user to just enable blosc or hdf5-blosc without enabling another.

ptim0626 replied:
One situation is if the user installed `blosc` in a custom path like `/my/path/to/blosc` (passed via `--with-blosc`), but `hdf5-blosc` is in the system path like `/usr/local`, and with `--enable-compression`, I think this is still a valid usage. Likewise for custom `hdf5-blosc` and system `blosc`.

ptim0626 commented:
Just browsing the `configure.ac` in hdf5 repo and they have [this check for silent rule](https://github.com/HDFGroup/hdf5/blob/hdf5_1.14.6/configure.ac#L36), we can follow suit.

teoching0705 committed:
Change the third argument to be issue page

teoching0705 committed:
change default debug to no

teoching0705 committed:
Modify configure.ac

change difference typos, change different steps, and checking HDF5 paths

teoching0705 committed:
change silent rules to use m4_ifdef

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,149 @@
+AS_IF([test -n "$HDFDIR"], [
+  # User gave --with-hdf5 — use it directly
+  CPPFLAGS="$CPPFLAGS -I$HDFDIR/include"
+  LDFLAGS="$LDFLAGS -L$HDFDIR/lib -Wl,-rpath=$HDFDIR/lib"
+
+  AC_CHECK_HEADER([hdf5.h], [], [AC_MSG_ERROR([Missing hdf5.h in $HDFDIR/include])])
+  AC_CHECK_LIB([hdf5], [H5open], [], [AC_MSG_ERROR([Missing libhdf5 in $HDFDIR/lib])])
+  AC_DEFINE([HAVE_HDF5], [1], [Define if HDF5 is available])
+],
+[
+  # No --with-hdf5: try different common path
+  AC_MSG_NOTICE([Check the path of $HDF5_ROOT])
+  AS_IF([test -n "$HDF5_ROOT"], [
+    CPPFLAGS="$CPPFLAGS -I$HDF5_ROOT/include"
+    LDFLAGS="$LDFLAGS -L$HDF5_ROOT/lib -Wl,-rpath=$HDF5_ROOT/lib"
+
+    AC_CHECK_HEADER([hdf5.h], [], [AC_MSG_ERROR([Missing hdf5.h in $HDF5_ROOT/include])])
+    AC_CHECK_LIB([hdf5], [H5open], [], [AC_MSG_ERROR([Missing libhdf5 in $HDF5_ROOT/lib])])
+    AC_DEFINE([HAVE_HDF5], [1], [Define if HDF5 is available])
+  ],
+  [
+    AC_MSG_NOTICE([Check the path of $HDF5_HOME])
+    AS_IF([test -n "$HDF5_HOME"], [
+      CPPFLAGS="$CPPFLAGS -I$HDF5_HOME/include"
+      LDFLAGS="$LDFLAGS -L$HDF5_HOME/lib -Wl,-rpath=$HDF5_HOME/lib"
+
+      AC_CHECK_HEADER([hdf5.h], [], [AC_MSG_ERROR([Missing hdf5.h in $HDF5_HOME/include])])
+      AC_CHECK_LIB([hdf5], [H5open], [], [AC_MSG_ERROR([Missing libhdf5 in $HDF5_HOME/lib])])
+      AC_DEFINE([HAVE_HDF5], [1], [Define if HDF5 is available])
+    ],
+    [
+      AC_MSG_NOTICE([Check the path of $HDF5_DIR])
+      AS_IF([test -n "$HDF5_DIR"], [
+        CPPFLAGS="$CPPFLAGS -I$HDF5_DIR/include"
+        LDFLAGS="$LDFLAGS -L$HDF5_DIR/lib -Wl,-rpath=$HDF5_DIR/lib"
+
+        AC_CHECK_HEADER([hdf5.h], [], [AC_MSG_ERROR([Missing hdf5.h in $HDF5_DIR/include])])
+        AC_CHECK_LIB([hdf5], [H5open], [], [AC_MSG_ERROR([Missing libhdf5 in $HDF5_DIR/lib])])
+        AC_DEFINE([HAVE_HDF5], [1], [Define if HDF5 is available])
+      ],
```
You can avoid the duplicated codes by assigning the corresponding environment variable to `HDFDIR` and perform only one go of `CPPFLAGS`/`LDFLAGS` and header/lib check.

teoching0705 committed:
Modify logic to remove redundant code

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,125 @@
+  [HBDIR="$withval"],
+  [])
+
+AS_IF([test "x$with_compression" != xno], [
```
Should this be `enable_compression` now? Same as line 106.

ptim0626 commented on configure.ac:
```diff
@@ -0,0 +1,125 @@
+  AS_IF([test -n "$BLOSCDIR" && test -n "$HBDIR"], [
+
+    CPPFLAGS="$CPPFLAGS -I$BLOSCDIR/include -I$HBDIR/src"
+    LDFLAGS="$LDFLAGS -L$BLOSCDIR/lib64 -L$HBDIR/build"
+    LDFLAGS="$LDFLAGS -Wl,-rpath=$BLOSCDIR/lib64 -Wl,-rpath=$HBDIR/build"
+
+    AC_CHECK_HEADER([blosc.h], [],
+      [AC_MSG_ERROR([blosc.h not found in $BLOSCDIR/include])])
+    AC_CHECK_LIB([blosc], [blosc_init], [],
+      [AC_MSG_ERROR([libblosc not found in $BLOSCDIR/lib64])])
+
+    AC_CHECK_HEADER([blosc_filter.h], [],
+      [AC_MSG_ERROR([blosc_filter.h not found in $HBDIR/src])])
+    AC_CHECK_LIB([blosc_filter], [register_blosc], [],
+      [AC_MSG_ERROR([libblosc_filter not found in $HBDIR/build])])
```
The directory `$HBDIR/build` looks like it isn't an installed one?

teoching0705 replied:
It is not.

from https://github.com/Blosc/hdf5-blosc?tab=readme-ov-file#compiling:

> Compiling
> 
> The filter consists of a single 'src/blosc_filter.c' source file and 'src/blosc_filter.h' header, which will need the Blosc library installed to work. It is simplest to just use the provided cmake build scripts, which compile and both the filter and the Blosc library into a library for you
> 
> Assuming you have [cmake](http://www.cmake.org/) and other standard Unix build tools installed, do:
> 
> mkdir build
> cd build
> cmake ..
> make
> 
> This generates the library/plugin files required above in the build directory.

ptim0626 replied:
What happens if you do `make install` afterwards (i.e. `cmake ..`, then `make`, then `make install`).

I remember something like `-DCMAKE_INSTALL_PREFIX` is how you specify the installation directory in cmake normally.

teoching0705 replied:
I am not sure, but this is the Makefile created by cmake:

```
# CMAKE generated file: DO NOT EDIT!
# Generated by "Unix Makefiles" Generator, CMake Version 3.26

# Default target executed when no arguments are given to make.
default_target: all
.PHONY : default_target

# Allow only one "make -f Makefile2" at a time, but pass parallelism.
.NOTPARALLEL:

#=============================================================================
# Special targets provided by cmake.

# Disable implicit rules so canonical targets will work.
.SUFFIXES:

# Disable VCS-based implicit rules.
% : %,v

# Disable VCS-based implicit rules.
% : RCS/%

# Disable VCS-based implicit rules.
% : RCS/%,v

# Disable VCS-based implicit rules.
% : SCCS/s.%

# Disable VCS-based implicit rules.
% : s.%

.SUFFIXES: .hpux_make_needs_suffix_list

# Command-line flag to silence nested $(MAKE).
$(VERBOSE)MAKESILENT = -s

#Suppress display of executed commands.
$(VERBOSE).SILENT:

# A target that is always out of date.
cmake_force:
.PHONY : cmake_force

#=============================================================================
# Set environment variables for the build.

# The shell in which to execute make rules.
SHELL = /bin/sh

# The CMake executable.
CMAKE_COMMAND = /usr/bin/cmake

# The command to remove a file.
RM = /usr/bin/cmake -E rm -f

# Escaping for special characters.
EQUALS = =

# The top-level source directory on which CMake was run.
CMAKE_SOURCE_DIR = /home/wck38436/Desktop/hdf5-blosc

# The top-level build directory on which CMake was run.
CMAKE_BINARY_DIR = /home/wck38436/Desktop/hdf5-blosc/build

#=============================================================================
# Targets provided globally by CMake.

# Special rule for the target test
test:
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Running tests..."
	/usr/bin/ctest --force-new-ctest-process $(ARGS)
.PHONY : test

# Special rule for the target test
test/fast: test
.PHONY : test/fast

# Special rule for the target edit_cache
edit_cache:
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Running CMake cache editor..."
	/usr/bin/ccmake -S$(CMAKE_SOURCE_DIR) -B$(CMAKE_BINARY_DIR)
.PHONY : edit_cache

# Special rule for the target edit_cache
edit_cache/fast: edit_cache
.PHONY : edit_cache/fast

# Special rule for the target rebuild_cache
rebuild_cache:
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Running CMake to regenerate build system..."
	/usr/bin/cmake --regenerate-during-build -S$(CMAKE_SOURCE_DIR) -B$(CMAKE_BINARY_DIR)
.PHONY : rebuild_cache

# Special rule for the target rebuild_cache
rebuild_cache/fast: rebuild_cache
.PHONY : rebuild_cache/fast

# Special rule for the target list_install_components
list_install_components:
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Available install components are: \"HDF5_FILTER_DEV\""
.PHONY : list_install_components

# Special rule for the target list_install_components
list_install_components/fast: list_install_components
.PHONY : list_install_components/fast

# Special rule for the target install
install: preinstall
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Install the project..."
	/usr/bin/cmake -P cmake_install.cmake
.PHONY : install

# Special rule for the target install
install/fast: preinstall/fast
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Install the project..."
	/usr/bin/cmake -P cmake_install.cmake
.PHONY : install/fast

# Special rule for the target install/local
install/local: preinstall
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Installing only the local directory..."
	/usr/bin/cmake -DCMAKE_INSTALL_LOCAL_ONLY=1 -P cmake_install.cmake
.PHONY : install/local

# Special rule for the target install/local
install/local/fast: preinstall/fast
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Installing only the local directory..."
	/usr/bin/cmake -DCMAKE_INSTALL_LOCAL_ONLY=1 -P cmake_install.cmake
.PHONY : install/local/fast

# Special rule for the target install/strip
install/strip: preinstall
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Installing the project stripped..."
	/usr/bin/cmake -DCMAKE_INSTALL_DO_STRIP=1 -P cmake_install.cmake
.PHONY : install/strip

# Special rule for the target install/strip
install/strip/fast: preinstall/fast
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Installing the project stripped..."
	/usr/bin/cmake -DCMAKE_INSTALL_DO_STRIP=1 -P cmake_install.cmake
.PHONY : install/strip/fast

# The main all target
all: cmake_check_build_system
	$(CMAKE_COMMAND) -E cmake_progress_start /home/wck38436/Desktop/hdf5-blosc/build/CMakeFiles /home/wck38436/Desktop/hdf5-blosc/build//CMakeFiles/progress.marks
	$(MAKE) $(MAKESILENT) -f CMakeFiles/Makefile2 all
	$(CMAKE_COMMAND) -E cmake_progress_start /home/wck38436/Desktop/hdf5-blosc/build/CMakeFiles 0
.PHONY : all

# The main clean target
clean:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/Makefile2 clean
.PHONY : clean

# The main clean target
clean/fast: clean
.PHONY : clean/fast

# Prepare targets for installation.
preinstall: all
	$(MAKE) $(MAKESILENT) -f CMakeFiles/Makefile2 preinstall
.PHONY : preinstall

# Prepare targets for installation.
preinstall/fast:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/Makefile2 preinstall
.PHONY : preinstall/fast

# clear depends
depend:
	$(CMAKE_COMMAND) -S$(CMAKE_SOURCE_DIR) -B$(CMAKE_BINARY_DIR) --check-build-system CMakeFiles/Makefile.cmake 1
.PHONY : depend

#=============================================================================
# Target rules for targets named project_blosc

# Build rule for target.
project_blosc: cmake_check_build_system
	$(MAKE) $(MAKESILENT) -f CMakeFiles/Makefile2 project_blosc
.PHONY : project_blosc

# fast build rule for target.
project_blosc/fast:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/project_blosc.dir/build.make CMakeFiles/project_blosc.dir/build
.PHONY : project_blosc/fast

#=============================================================================
# Target rules for targets named blosc_filter_shared

# Build rule for target.
blosc_filter_shared: cmake_check_build_system
	$(MAKE) $(MAKESILENT) -f CMakeFiles/Makefile2 blosc_filter_shared
.PHONY : blosc_filter_shared

# fast build rule for target.
blosc_filter_shared/fast:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/blosc_filter_shared.dir/build.make CMakeFiles/blosc_filter_shared.dir/build
.PHONY : blosc_filter_shared/fast

#=============================================================================
# Target rules for targets named blosc_plugin_shared

# Build rule for target.
blosc_plugin_shared: cmake_check_build_system
	$(MAKE) $(MAKESILENT) -f CMakeFiles/Makefile2 blosc_plugin_shared
.PHONY : blosc_plugin_shared

# fast build rule for target.
blosc_plugin_shared/fast:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/blosc_plugin_shared.dir/build.make CMakeFiles/blosc_plugin_shared.dir/build
.PHONY : blosc_plugin_shared/fast

#=============================================================================
# Target rules for targets named example

# Build rule for target.
example: cmake_check_build_system
	$(MAKE) $(MAKESILENT) -f CMakeFiles/Makefile2 example
.PHONY : example

# fast build rule for target.
example/fast:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/example.dir/build.make CMakeFiles/example.dir/build
.PHONY : example/fast

src/blosc_filter.o: src/blosc_filter.c.o
.PHONY : src/blosc_filter.o

# target to build an object file
src/blosc_filter.c.o:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/blosc_filter_shared.dir/build.make CMakeFiles/blosc_filter_shared.dir/src/blosc_filter.c.o
	$(MAKE) $(MAKESILENT) -f CMakeFiles/blosc_plugin_shared.dir/build.make CMakeFiles/blosc_plugin_shared.dir/src/blosc_filter.c.o
.PHONY : src/blosc_filter.c.o

src/blosc_filter.i: src/blosc_filter.c.i
.PHONY : src/blosc_filter.i

# target to preprocess a source file
src/blosc_filter.c.i:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/blosc_filter_shared.dir/build.make CMakeFiles/blosc_filter_shared.dir/src/blosc_filter.c.i
	$(MAKE) $(MAKESILENT) -f CMakeFiles/blosc_plugin_shared.dir/build.make CMakeFiles/blosc_plugin_shared.dir/src/blosc_filter.c.i
.PHONY : src/blosc_filter.c.i

src/blosc_filter.s: src/blosc_filter.c.s
.PHONY : src/blosc_filter.s

# target to generate assembly for a file
src/blosc_filter.c.s:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/blosc_filter_shared.dir/build.make CMakeFiles/blosc_filter_shared.dir/src/blosc_filter.c.s
	$(MAKE) $(MAKESILENT) -f CMakeFiles/blosc_plugin_shared.dir/build.make CMakeFiles/blosc_plugin_shared.dir/src/blosc_filter.c.s
.PHONY : src/blosc_filter.c.s

src/blosc_plugin.o: src/blosc_plugin.c.o
.PHONY : src/blosc_plugin.o

# target to build an object file
src/blosc_plugin.c.o:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/blosc_plugin_shared.dir/build.make CMakeFiles/blosc_plugin_shared.dir/src/blosc_plugin.c.o
.PHONY : src/blosc_plugin.c.o

src/blosc_plugin.i: src/blosc_plugin.c.i
.PHONY : src/blosc_plugin.i

# target to preprocess a source file
src/blosc_plugin.c.i:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/blosc_plugin_shared.dir/build.make CMakeFiles/blosc_plugin_shared.dir/src/blosc_plugin.c.i
.PHONY : src/blosc_plugin.c.i

src/blosc_plugin.s: src/blosc_plugin.c.s
.PHONY : src/blosc_plugin.s

# target to generate assembly for a file
src/blosc_plugin.c.s:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/blosc_plugin_shared.dir/build.make CMakeFiles/blosc_plugin_shared.dir/src/blosc_plugin.c.s
.PHONY : src/blosc_plugin.c.s

src/example.o: src/example.c.o
.PHONY : src/example.o

# target to build an object file
src/example.c.o:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/example.dir/build.make CMakeFiles/example.dir/src/example.c.o
.PHONY : src/example.c.o

src/example.i: src/example.c.i
.PHONY : src/example.i

# target to preprocess a source file
src/example.c.i:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/example.dir/build.make CMakeFiles/example.dir/src/example.c.i
.PHONY : src/example.c.i

src/example.s: src/example.c.s
.PHONY : src/example.s

# target to generate assembly for a file
src/example.c.s:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/example.dir/build.make CMakeFiles/example.dir/src/example.c.s
.PHONY : src/example.c.s

# Help Target
help:
	@echo "The following are some of the valid targets for this Makefile:"
	@echo "... all (the default if no target is provided)"
	@echo "... clean"
	@echo "... depend"
	@echo "... edit_cache"
	@echo "... install"
	@echo "... install/local"
	@echo "... install/strip"
	@echo "... list_install_components"
	@echo "... rebuild_cache"
	@echo "... test"
	@echo "... project_blosc"
	@echo "... blosc_filter_shared"
	@echo "... blosc_plugin_shared"
	@echo "... example"
	@echo "... src/blosc_filter.o"
	@echo "... src/blosc_filter.i"
	@echo "... src/blosc_filter.s"
	@echo "... src/blosc_plugin.o"
	@echo "... src/blosc_plugin.i"
	@echo "... src/blosc_plugin.s"
	@echo "... src/example.o"
	@echo "... src/example.i"
	@echo "... src/example.s"
.PHONY : help



#=============================================================================
# Special targets to cleanup operation of make.

# Special rule to run CMake to check the build system integrity.
# No rule that depends on this can have commands that come from listfiles
# because they might be regenerated.
cmake_check_build_system:
	$(CMAKE_COMMAND) -S$(CMAKE_SOURCE_DIR) -B$(CMAKE_BINARY_DIR) --check-build-system CMakeFiles/Makefile.cmake 0
.PHONY : cmake_check_build_system

```

ptim0626 replied:
If you do
```bash
mkdir build
cd build
cmake -DCMAKE_INSTALL_PREFIX=/your/path/to/h5b -DPLUGIN_INSTALL_PATH=/your/path/to/h5-plugin ..
make -j4
make install
```
will install things into their respective directory.

The `build` directory is typically just for the building process, people will often `make install` afterwards to put the libraries/headers to the prefix or the default location.

teoching0705 committed:
Change with-compression to enable-compression

teoching0705 committed:
Fix hdf5 not correctly detected when HDFDIR alredy set before the configuration

teoching0705 committed:
Separate blosc and hdf5-blosc detection

ptim0626 commented:
After removing those GNU files and dealing with the build directory in hdf5-blosc, I think we are good to go.

teoching0705 committed:
Modify checking logics of HDF5 to make it more robust

teoching0705 committed:
modify logics to check compression dependencies

teoching0705 committed:
Remove GNU files

ptim0626 commented:
1. You will need to rebase with the `dev` branch. (the src includes `hdf5_init.c` etc now)
2. Remove the `.jython_cache` directory (accidentally committed?)
3. The installation directory of hdf5-blosc should not contain `build`, the directory structure looks like this if you do `make install`:
```bash
h5b-installed/
├── include
│   └── blosc_filter.h
└── lib
    └── libblosc_filter.so
```
4. It looks like the installation of `hdf5-blosc` contains `blosc` as well. Can we use that instead of installing a separate one? We may be able to get away with one dependency.

teoching0705 committed:
Remove jython files

teoching0705 commented:
> 1. You will need to rebase with the `dev` branch. (the src includes `hdf5_init.c` etc now)
> 
>     2. Remove the `.jython_cache` directory (accidentally committed?)
> 
>     3. The installation directory of hdf5-blosc should not contain `build`, the directory structure looks like this if you do `make install`:
> 
> 
> ```shell
> h5b-installed/
> ├── include
> │   └── blosc_filter.h
> └── lib
>     └── libblosc_filter.so
> ```
> 
>     4. It looks like the installation of `hdf5-blosc` contains `blosc` as well. Can we use that instead of installing a separate one? We may be able to get away with one dependency.

I think the hdf5-blosc installed blosc on the build directory that the user create when following the cmake instruction. I think the user can provide a path to the hdf5-blosc/build/blosc for the blosc.

Commit:
change directory for checking hdf5-blosc
